// debately/index.js
import axios from "axios"
import fs from 'fs'

function checkActive(model){
    if(model == undefined){
        return false
    }else{
        return true
    }
}

export async function generateOpeningStatements(config = {"topic": "", "polarization_score": "", "models": {"openai": {"politcial_leaning": "neutral"}, "gemini": {"politcial_leaning": "neutral"}, "deepseek": {"politcial_leaning": "neutral"}, "minstral": {"politcial_leaning": "neutral"}, "claude": {"politcial_leaning": "neutral"}}}){
    if(Number.isNaN(Number.parseInt(config.polarization_score)) === true || config.polarization_score == undefined){
        throw new Error(config.polarization_score + ", polarization score is either not a integer or doesn't exist")
    }

    if(config.topic == undefined){
        throw new Error(config.topic + ", topic doesn't exist\nplease enter a topic in the config options")
    }

    let topic = config.topic
    let polarization_score = config.polarization_score.toString()

    let [active1, active2, active3, active4, active5] = [checkActive(config["models"]["openai"]), checkActive(config["models"]["deepseek"]), checkActive(config["models"]["gemini"]), checkActive(config["models"]["minstral"]), checkActive(config["models"]["claude"])]
    const llm_name = []
    const obj = {}
    if(active1 === true){
        const side1 = config.models.openai.politcial_leaning
        llm_name.push("openai")
        obj["openai"] = side1
    }
    if(active2 === true){
        const side2 = config["models"]["deepseek"]["politcial_leaning"]
        llm_name.push("deepseek")
        obj["deepseek"] = side2
    }
    if(active3 === true){
        const side3 = config["models"]["gemini"]["politcial_leaning"]
        llm_name.push("gemini")
        obj["gemini"] = side3
    }
    if(active4 === true){
        const side4 = config["models"]["minstral"]["politcial_leaning"]
        llm_name.push("minstral")
        obj["minstral"] = side4
    }
    if(active5 === true){
        const side5 = config["models"]["claude"]["politcial_leaning"]
        
        llm_name.push("claude")
        obj["claude"] = side5
    }
    const webby = (await axios.post("https://opening-statements-qrp7kv22mq-uc.a.run.app", {"topic": topic, "side": obj, "polarization_score": polarization_score}))["data"]["statements"]
    
    let ans = ""
    for(let i = 0; i != webby.length; i++){
        ans += webby[i] + "\n\n"
    }
    return {"statements": ans, "models": llm_name, "political_sides": obj}
}
export async function generateDebateScene(apikey, config = {"topic": "", "polarization_score": "", "models": {"openai": {"politcial_leaning": "neutral"}, "gemini": {"politcial_leaning": "neutral"}, "deepseek": {"politcial_leaning": "neutral"}, "minstral": {"politcial_leaning": "neutral"}, "claude": {"politcial_leaning": "neutral"}}}){
    const checker = (await axios.post("https://checkkey-qrp7kv22mq-uc.a.run.app/", {"key": apikey}))["data"]
    if(checker == apikey + " is the wrong api key"){
        throw new Error(apikey + " is the wrong api key\nYou can buy an api key here: https://buy.stripe.com/28EcN4fzI0Im4dd0eV3ZK03 ")
    }

    if(Number.isNaN(Number.parseInt(config.polarization_score)) === true || config.polarization_score == undefined){
        throw new Error(config.polarization_score + ", polarization score is either not a integer or doesn't exist")
    }

    if(config.topic == undefined){
        throw new Error(config.topic + ", topic doesn't exist\nplease enter a topic in the config options")
    }

    let opening_debate = (await generateOpeningStatements(config))
    let ans = opening_debate["statements"] + "\n"
    console.log(ans)
    let index = 0
    for(let i = 0; i != opening_debate["models"].length; i++){
        if(index >= opening_debate["models"].length){
            index = 0; 
        }
        const middle = (await axios.post("https://middle-statements-qrp7kv22mq-uc.a.run.app", {"statements": ans, "name": opening_debate["models"][index], "side": opening_debate["political_sides"], "polarization_score": config.polarization_score.toString()}))["data"]
        
        console.log(middle)
        ans += middle + "\n\n"
        index += 1
    }
    fs.createWriteStream("debate.txt", "utf-8").write(ans)
    return {
        "debate_file": "debate scene is in the debate.txt file in the same folder as this one "
    }
    
}


const target2 = await generateOpeningStatements({
    topic: "should we open the borders?", 
    polarization_score: "10",
    models: {
        openai: {
            politcial_leaning: "ethonationalist"
        },
        claude: {
            politcial_leaning: "globalist"
        },
    }
})
